### PR TITLE
fix DescribeTargetGroupTargetsForArn with empty target list

### PR DESCRIPTION
**Pull request by r0fls**
fixes: #383

### CODE REVIEW

**Comment by bigkraig** [comment]
I’m operating off a phone in Costa Rica right now. So my review is limited, but I’d like to get this fixed for you guys. Is there a reason why we only look for len==1 and not join the lists whenever it’s => 1? And if a third party can confirm this fixes it for them I’m good with a phone merge. I can’t hover to see who 👍

**Comment by r0fls** [comment]
@bigkraig 😂 thanks for checking this out while you're on vacation. Jesus and Khris form the 👍 party

**Comment by jkemp101** [comment]
This didn't seem to fix it for me.  Did it work for you guys?  Can we add a test for this situation to catch issues like this in the future?

**Comment by jkemp101** [comment]
The newly added instance is falling into `case elbv2.TargetHealthStateEnumUnused:` and therefore not being added to `result`.

**Comment by khrisrichardson** [comment]
@jkemp101 check out the reasons for state unused [here](https://docs.aws.amazon.com/sdk-for-go/api/service/elbv2/#TargetHealth). I don't believe the target should be added in any of those circumstances, but am willing to hear arguments to the contrary.

```
    // If the target state is unused, the reason code can be one of the following
    // values:
    //
    //    * Target.NotRegistered - The target is not registered with the target
    //    group.
    //
    //    * Target.NotInUse - The target group is not used by any load balancer
    //    or the target is in an Availability Zone that is not enabled for its load
    //    balancer.
    //
    //    * Target.IpUnusable - The target IP address is reserved for use by a load
    //    balancer.
    //
    //    * Target.InvalidState - The target is in the stopped or terminated state.
```

**Comment by jkemp101** [comment]
I assume it is hitting that state because `Target.NotRegistered` which would make sense because it is a new instance that hasn't been added to the target group yet but that is what is needed to be done.  Can you confirm if this is working for you or anyone?  AWS instances fail all of the time.  As far as I can tell if an instance fails and is replaced, the new one will not get added to the target group.  Eventually you would be left with no instances in the target group and your site would go down.

I'm going back to CLBs via regular k8s LoadBalancer services for now because of this issue and my other concerns laid out in https://github.com/coreos/alb-ingress-controller/issues/375.  I need to move forward with other parts of my project and will revisit k8s ALB options later.

**Comment by roffe** [comment]
PR fixed my problem with ALB's loosing all the targets when i do a kops rolling upgrade on the cluster and replace all nodes

File: pkg/aws/elbv2/elbv2.go [hunk 1]
```diff
@@ -207,20 +207,13 @@ func (e *ELBV2) DescribeTagsForArn(arn *string) (util.Tags, error) {
...
-		})
-	} else {
-		targetHealth, err = e.DescribeTargetHealth(&elbv2.DescribeTargetHealthInput{
-			TargetGroupArn: arn,
-		})
+	opts := &elbv2.DescribeTargetHealthInput{
+		TargetGroupArn: arn,
+	}
+	for _, targets := range targets {
+		mergedTargets = append(opts.Targets, target...)
```
Comment (khrisrichardson) [inline pkg/aws/elbv2/elbv2.go]: I think you mean `opts.Targets = append....`

Comment (r0fls) [reply pkg/aws/elbv2/elbv2.go]: Whoops! ☕️

Comment (r0fls) [reply pkg/aws/elbv2/elbv2.go]: Bah, I did a `go build cmd` but I guess that wouldn't have caught it anyway...

**Pull request merged**